leo_filters/src/imu_filter.cpp

```diff
@@ -0,0 +1,267 @@
+
+  imu_sub_ = create_subscription<sensor_msgs::msg::Imu>(
+    "imu/data_raw", rclcpp::QoS(5).best_effort(),
+    std::bind(&ImuFilter::imu_callback, this, _1));
+
+  RCLCPP_INFO(get_logger(), "Started node");
+}
+
+void ImuFilter::save_bias()
+{
+  YAML::Node node;
+  std::string file_path = get_bias_file_path();
+
+  try {
+    node = YAML::LoadFile(file_path);
+    node["gyro_bias_x"] = filter_.getAngularVelocityBiasX();
+    node["gyro_bias_y"] = filter_.getAngularVelocityBiasY();
+    node["gyro_bias_z"] = filter_.getAngularVelocityBiasZ();
+  } catch (const YAML::BadFile & e) {
+    RCLCPP_ERROR(
+      get_logger(),
+      "IMU bias file doesn't exist or couldn't be opened: %s",
+      e.what());
+  } catch (const YAML::Exception & e) {
+    RCLCPP_ERROR(
+      get_logger(), "YAML error while loading bias file: %s",
+      e.what());
+    return;
+  }
+
+  try {
+    std::ofstream fout(file_path);
+    fout << node;
+
+    if (!fout.is_open()) {
+      RCLCPP_ERROR(
+        get_logger(), "Failed to open bias file for writing: %s",
+        file_path.c_str());
+      return;
+    }
+  } catch (const YAML::Exception & e) {
+    RCLCPP_ERROR(
+      get_logger(), "YAML error while writing bias file: %s",
+      e.what());
+  }
+}
+
+void ImuFilter::load_bias()
+{
+  YAML::Node node;
+  std::string file_path = get_bias_file_path();
+
+  try {
+    node = YAML::LoadFile(file_path);
+
+    if (node["gyro_bias_x"]) {
+      filter_.setAngularVelocityBiasX(node["gyro_bias_x"].as<double>());
+    }
+
+    if (node["gyro_bias_y"]) {
+      filter_.setAngularVelocityBiasY(node["gyro_bias_y"].as<double>());
+    }
+
+    if (node["gyro_bias_z"]) {
+      filter_.setAngularVelocityBiasZ(node["gyro_bias_z"].as<double>());
+    }
+
+  } catch (YAML::BadFile & e) {
+    RCLCPP_ERROR(get_logger(), "IMU bias file doesn't exist.\n");
+    RCLCPP_ERROR(get_logger(), "Creating IMU bias file with current gyrometer bias.\n");
+
+    node["gyro_bias_x"] = filter_.getAngularVelocityBiasX();
+    node["gyro_bias_y"] = filter_.getAngularVelocityBiasY();
+    node["gyro_bias_z"] = filter_.getAngularVelocityBiasZ();
+
+    std::ofstream fout(file_path);
+    fout << node;
+  }
+}
+
+std::string ImuFilter::get_bias_file_path()
+{
+  std::string ros_home;
+  char * ros_home_env;
+  if (ros_home_env = std::getenv("ROS_HOME")) {
+    ros_home = ros_home_env;
+  } else if (ros_home_env = std::getenv("HOME")) {
+    ros_home = ros_home_env;
+    ros_home += "/.ros";
+  }
+
+  return ros_home + "/imu_calibration.yaml";
+}
+
+void ImuFilter::update_filter_params()
+{
+  if (filter_.getDoBiasEstimation() != params_.do_bias_estimation) {
+    filter_.setDoBiasEstimation(params_.do_bias_estimation);
+  }
+
+  if (filter_.getDoAdaptiveGain() != params_.do_adaptive_gain) {
+    filter_.setDoAdaptiveGain(params_.do_adaptive_gain);
+  }
+
+  double gain = filter_.getGainAcc();
+  if (std::fabs(gain - params_.gain_acc) > 0.001) {
```

Comment:
Why not just check if its different?
```suggestion
  if (gain != params_.gain_acc) {
```